driven/storage/adapter_surveys.go

```diff
@@ -219,65 +223,110 @@ func (a *Adapter) GetSurveysWithResponses(orgID string, appID string, userID *st
+	keepResponses := includeResponses != nil && *includeResponses
+	needLookup := userIDStr != "" && (resultDependsOnResponses || keepResponses)
+	if needLookup {
+		// Build lookup pipeline and add $limit: 1 only when includeResponses is false
+		responseLookupPipeline := bson.A{
+			bson.D{{Key: "$match", Value: bson.M{
+				"$expr": bson.M{
+					"$and": bson.A{
+						bson.M{"$eq": bson.A{"$survey._id", "$$surveyIdVar"}},
+						bson.M{"$eq": bson.A{"$org_id", orgID}},
+						bson.M{"$eq": bson.A{"$app_id", appID}},
+						bson.M{"$eq": bson.A{"$user_id", userIDStr}},
 					},
-				}}},
-				bson.D{{Key: "$project", Value: bson.D{
-					{Key: "_id", Value: 1},
-					{Key: "user_id", Value: 1},
-					{Key: "date_created", Value: 1},
-					{Key: "survey", Value: 1},
-				}}},
+				},
+			}}},
+		}
+		if !keepResponses {
+			responseLookupPipeline = append(responseLookupPipeline, bson.D{{Key: "$limit", Value: 1}})
+		}
+		responseLookupPipeline = append(responseLookupPipeline, bson.D{{Key: "$project", Value: bson.D{
+			{Key: "_id", Value: 1},
+			{Key: "user_id", Value: 1},
+			{Key: "date_created", Value: 1},
+			{Key: "survey", Value: 1},
+		}}})
+
+		// Lookup survey_responses for each survey matching the survey_id and user_id.
+		lookupStage := bson.D{
+			{Key: "$lookup", Value: bson.D{
+				{Key: "from", Value: "survey_responses"},
+				// `surveyIdVar` holds the value of the survey's _id
+				{Key: "let", Value: bson.D{{Key: "surveyIdVar", Value: "$_id"}}},
+				{Key: "pipeline", Value: responseLookupPipeline},
+				{Key: "as", Value: "survey_responses"},
 			}},
-			{Key: "as", Value: "survey_responses"},
-		}},
-	}
-	pipeline = append(pipeline, lookupStage)
-
-	// Compute a boolean field "Completed" based on whether a response exists
-	addCompletedFieldStage := bson.D{
-		{Key: "$addFields", Value: bson.D{
-			{Key: "completed", Value: bson.D{
-				{Key: "$cond", Value: bson.A{
-					bson.D{{Key: "$gt", Value: bson.A{
-						bson.D{{Key: "$size", Value: "$survey_responses"}},
-						0,
-					}}},
-					true,
-					false,
+		}
+		pipeline = append(pipeline, lookupStage)
+
+		// Compute a boolean field "Completed" based on whether a response exists
+		addCompletedFieldStage := bson.D{
+			{Key: "$addFields", Value: bson.D{
+				{Key: "completed", Value: bson.D{
+					{Key: "$cond", Value: bson.A{
+						bson.D{{Key: "$gt", Value: bson.A{
+							bson.D{{Key: "$size", Value: "$survey_responses"}},
+							0,
+						}}},
+						true,
+						false,
+					}},
 				}},
 			}},
-		}},
-	}
-	pipeline = append(pipeline, addCompletedFieldStage)
+		}
+		pipeline = append(pipeline, addCompletedFieldStage)
 
-	// Optionally remove the joined survey responses from the output.
-	if includeResponses != nil && (!*includeResponses) {
-		projectStage := bson.D{
-			{Key: "$project", Value: bson.D{
-				{Key: "survey_responses", Value: 0},
-			}},
+		// Optionally remove the joined survey responses from the output.
+		if !keepResponses {
```

Comment:
It looks like the logic here changed from `includeResponses != nil && (!*includeResponses)` to `!(includeResponses != nil && *includeResponses)`. I think the impact of this is that when `includeResponses` is nil the responses will now be removed when they previously would not have been.

Reply:
Yes, you're right. I was setting `keepResponses` incorrectly.